### PR TITLE
Exportable kimi

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47096)

# What does this PR do?

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [ ] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/exporters/exporter_dynamo.py [hunk 1]
```diff
@@ -350,8 +358,12 @@ def _to_batched(t):
...
     )
 
-    # (n_seg, heads, seg_len, dim) → (n_seg, seg_len, heads, dim) → (seq, heads*dim)
-    attn_output = attn_output.transpose(1, 2).reshape(seq_length, -1).contiguous()
+    # (n_seg, heads, seg_len, dim) → (n_seg, seg_len, heads, dim) → (seq, heads*dim).
+    # The `.clone()` materialises the post-transpose tensor so the reshape stays a valid view:
+    # executorch's edge-lowering reshape ref refuses to copy a non-contiguous tensor, and a plain
+    # `.contiguous()` gets folded away by functionalization (a `.clone()` survives). Only bites
+    # multi-segment grids (n_seg > 1); single-image encoders reshape a size-1 leading dim.
+    attn_output = attn_output.transpose(1, 2).clone().reshape(seq_length, -1)
```
Comment (IlyasMoutawwakil) [inline src/transformers/exporters/exporter_dynamo.py]: should probably do this for ET only

File: src/transformers/exporters/exporter_dynamo.py [hunk 2]
```diff
@@ -350,8 +350,8 @@ def _to_batched(t):
         enable_gqa=getattr(self, "num_key_value_heads", self.num_heads) != self.num_heads,
     )
 
-    # (n_seg, heads, seg_len, dim) → (n_seg, seg_len, heads, dim) → (seq, heads*dim)
-    attn_output = attn_output.transpose(1, 2).reshape(seq_length, -1).contiguous()
+    # (n_seg, heads, seg_len, dim) → (n_seg, seg_len, heads, dim) → (seq, heads*dim).
+    attn_output = attn_output.transpose(1, 2).reshape(seq_length, -1)
```
Comment (vasqu) [inline src/transformers/exporters/exporter_dynamo.py:355]: hmm why remove?

File: src/transformers/exporters/exporter_executorch.py [hunk 3]
```diff
@@ -418,6 +418,25 @@ def patch(self, *sizes):
     return patch
 
 
+@register_patch("executorch", "torch.reshape", "torch.Tensor.reshape")
+def _patch_reshape(original):
```
Comment (vasqu) [inline src/transformers/exporters/exporter_executorch.py:431]: hmm ok so here we have the issue indirectly re contiguous?

File: src/transformers/exporters/utils.py [hunk 4]
```diff
@@ -684,9 +708,13 @@ def _find_multimodal_submodules(model: PreTrainedModel) -> dict[str, torch.nn.Mo
...
-def is_multimodal(model: PreTrainedModel) -> bool:
-    """Returns `True` if the model is multi-modal with modal encoders and a language model."""
-    return bool(_find_multimodal_submodules(model))
+def is_multimodal(model: PreTrainedModel | torch.nn.Module) -> bool:
+    """Returns `True` if the model is multi-modal with modal encoders and a language model.
+
+    A non-`PreTrainedModel` (e.g. a bare `nn.Module`) has no canonical `get_encoder`/`get_decoder`
+    accessors and is trivially not multi-modal, so it short-circuits to `False`.
+    """
+    return isinstance(model, PreTrainedModel) and bool(_find_multimodal_submodules(model))
```
Comment (vasqu) [inline src/transformers/exporters/utils.py:727]: just fmi where do we suddenly call on a normal nn module?

File: src/transformers/models/kimi_k25/modular_kimi_k25.py [hunk 5]
```diff
@@ -157,35 +255,22 @@ def compute_pos_embed(self):
...
+        bicubic_indices, bicubic_weights = get_vision_bicubic_indices_and_weights(
+            grid_thw, self.num_grid_per_side, kwargs=kwargs
+        )
+        pos = F.embedding_bag(bicubic_indices, table, per_sample_weights=bicubic_weights.to(table.dtype), mode="sum")
+        # Temporal: add a per-frame sinusoid. Row 0 of the table is a zero pad, so single-frame clips
+        # (frame index 0) get none.
+        time_table = torch.cat(
+            [self.time_position_embeddings.new_zeros(1, self.dim), self.time_position_embeddings.squeeze(1)]
+        )
+        pos = pos + time_table[get_vision_frame_index(grid_thw, kwargs=kwargs)]
```
Comment (vasqu) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py]: is this not quite expensive?

File: src/transformers/models/kimi_k25/modular_kimi_k25.py [hunk 6]
```diff
@@ -197,27 +282,24 @@ def __init__(self, config):
...
+        hidden_states = self.pos_emb(hidden_states, grid_thw, **kwargs)
         return hidden_states
 
 
 # Similarly to gemma4, applies the same freq to H and W grids
 # The difference is that gemma4 stacks H/W embeds on `dim`, while Kimi interleaves them
 class Kimi_K25VisionRotaryEmbedding(Gemma4VisionRotaryEmbedding):
     def forward(self, x, position_ids):
-        position_ids_expanded = position_ids.permute(1, 2, 0)[..., None].float()  # shape (bs, positions, 2, 1)
+        position_ids_expanded = position_ids.transpose(0, 1)[..., None].float()  # (positions, 2, 1)
```
Comment (vasqu) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py:253]: what happened to the batch dim?

**Review by vasqu — COMMENTED** [review]
Just have a few smaller comments because the changes are honestly a bit hard to verify without integration tests cc @zucchini-nlp whenever you have time to check

File: src/transformers/exporters/exporter_dynamo.py (hunk 2 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/exporter_dynamo.py:355]: not needed in the default (dynamo export) path, if a backend needs it it should handle it on its own (see patch_reshape)

File: src/transformers/exporters/exporter_executorch.py (hunk 3 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/exporter_executorch.py:431]: the issue was always in executorch, imo this is more direct 😆

**Comment by zucchini-nlp** [comment]
I'm off 9-15 July so will do it after I am back (unless someone wants to add tests before that) 😄

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 6 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py:253]: the extra (batch) dim was always size 1 in this packed path, so carrying it through rope was pointless and it was squeezed out

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: hmm yes could be, it's part of vectorizing the old per-image grid_thw.tolist() loop (untraceable for export)
the zero-pad table replaced the old if t > 1 branch. you are right we can avoid rebuilding it per forward ! 
moved the pad into the buffer at init, so forward is now just a gather.

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25

File: src/transformers/exporters/utils.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/utils.py:727]: not suddenly tbh, this happens during testing and i didn't catch it before

File: src/transformers/exporters/utils.py [hunk 7]
```diff
@@ -444,9 +445,13 @@ def _prepare_grid_thw_vision_inputs(model: torch.nn.Module, inputs: dict[str, An
...
-    cu_seqlens, max_seqlen = get_vision_attention_seqlens(grid_thw, model.config, kwargs=inputs)
-    inputs["cu_seqlens"] = cu_seqlens
-    inputs["max_seqlen"] = max_seqlen
+    # kimi_k25-style encoders define their own per-frame / temporal-merge precompute helpers in their
+    # modeling module (resolved below) and attend over the whole clip, so `cu_seqlens` is per-clip
+    # (matching the encoder's util call). Other grid_thw encoders lack these and stay per-frame.
+    module = sys.modules[type(model).__module__]
+    temporal_encoder = hasattr(module, "get_vision_frame_index")
+    inputs["cu_seqlens"] = get_vision_cu_seqlens(grid_thw, merge_temporal=temporal_encoder)
+    inputs["max_seqlen"] = get_max_seqlen(inputs["cu_seqlens"], model.config, kwargs=inputs)
```
Comment (vasqu) [inline src/transformers/exporters/utils.py]: Why not use get_vision_attention_seqlens as before?

File: src/transformers/models/kimi_k25/modular_kimi_k25.py [hunk 8]
```diff
@@ -350,51 +430,6 @@ def __init__(self, config: Kimi_K25VisionConfig):
         self.final_layernorm = nn.LayerNorm(config.hidden_size, eps=1e-05)
         self.post_init()
 
-    def temporal_patch_merger(
```
Comment (vasqu) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py:353]: Technically slightly breaking can we keep BC for a bit at least?

File: src/transformers/models/kimi_k25/modular_kimi_k25.py [hunk 9]
```diff
@@ -407,19 +442,12 @@ def forward(
...
 
-        lengths = torch.cat(
-            (
-                torch.zeros(1, dtype=grid_thw.dtype, device=grid_thw.device),
-                grid_thw[:, 0] * grid_thw[:, 1] * grid_thw[:, 2],
-            )
-        )
-
-        cu_seqlens = lengths.cumsum(dim=0, dtype=torch.int32)
+        cu_seqlens = get_vision_cu_seqlens(grid_thw, merge_temporal=True, kwargs=kwargs)
```
Comment (vasqu) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py]: Can apply our fused version for max seq len together

File: src/transformers/models/kimi_k25/modular_kimi_k25.py
Comment (vasqu) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py:1]: General comment and not on you: But we do introduce some very model specific workarounds. I feel like it would make more and more sense to start moving things into a collator / processor that handles these

**Review by vasqu — APPROVED** [review]
Just some smaller comments but overll looks good

**Comment by vasqu** [comment]
Also cc @zucchini-nlp if you want to take a look

File: src/transformers/models/kimi_k25/modular_kimi_k25.py [hunk 10]
```diff
@@ -54,6 +54,101 @@
 logger = logging.get_logger(__name__)
 
 
+def get_vision_bicubic_indices_and_weights(
```
Comment (zucchini-nlp) [inline src/transformers/models/kimi_k25/modular_kimi_k25.py]: why we don't put it in `vision_utils`? we could let the interpolation accept mode arg, and I am also adding `align_coreners` 
I suppose we'll keep getting different interpolation modes in vision anyway

**Review by zucchini-nlp — APPROVED** [review]

File: src/transformers/exporters/utils.py (hunk 7 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/utils.py]: yes sure

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 8 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py:353]: reverted

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 9 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: yess nice catch

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 10 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: moved it to vision_utils, generally i put utils there once they are used by multiple models, while it's only used by one it can be subject to modifications in order to be used by others

Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: > we could let the interpolation accept mode arg

as in create a unifying utils `get_vision_interpolation_indices_and_weights` that accepts a mode argument ?

Comment (zucchini-nlp) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: yep :)

Imo bilinear and bicubic are the most common ones used in ViT models, so we can support both by default

File: src/transformers/models/kimi_k25/modular_kimi_k25.py
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py:1]: > But we do introduce some very model specific workarounds.

i wouldn't call it a workaround.. my understanding is that the collator / processor does not remove the need for this kind of rewrite (or write from day 0 preferably), it just makes it more useful and instrumentalize it to avoid graph breaks inside the model from dynamic tensors computation (cuz the utils are still data-dep after all). There's nothing to move into a collator / processor since these utils have to be called when the dynamic tensors are not passed.

File: src/transformers/models/kimi_k25/modular_kimi_k25.py (hunk 10 above)
Comment (IlyasMoutawwakil) [reply src/transformers/models/kimi_k25/modular_kimi_k25.py]: added a unified util, tell me what u think

File: src/transformers/exporters/utils.py [hunk 11]
```diff
@@ -470,15 +470,15 @@ def _prepare_grid_thw_vision_inputs(model: torch.nn.Module, inputs: dict[str, An
...
     if num_grid_per_side is not None:
-        if hasattr(module, "get_vision_bicubic_indices_and_weights"):
-            # kimi_k25 resamples its learned grid bicubically (its module imports the shared helper).
-            inputs["bicubic_indices"], inputs["bicubic_weights"] = module.get_vision_bicubic_indices_and_weights(
-                grid_thw, num_grid_per_side
-            )
+        # kimi_k25 (the temporal encoder) resamples its learned grid bicubically with no spatial-merge
+        # reorder; the other grid_thw encoders (qwen3_vl / qwen3_5 / paddleocr_vl) use bilinear.
+        if temporal_encoder:
+            mode, align_corners, interp_merge = "bicubic", False, 1
```
Comment (zucchini-nlp) [inline src/transformers/exporters/utils.py]: hmm, would work for now but we'd prob want to get this more reliably from module. Maybe each module would need to have `self.interpolation_mode` or smth

File: src/transformers/vision_utils.py [hunk 12]
```diff
@@ -184,126 +188,130 @@ def get_vision_window_index(
...
+    """Per-patch gather indices/weights that resample a square learned `(num_grid_per_side,
+    num_grid_per_side)` position-embedding table to each image's `(h, w)` grid, or pop
+    `"interp_indices"`/`"interp_weights"` from `kwargs` if both precomputed.
+
+    Reproduces `F.interpolate(mode=mode, align_corners=align_corners)` as `(total_patches, n_taps)`
+    indices + weights (`n_taps` = 4 for `"bilinear"`, 16 for `"bicubic"` — the Keys kernel with
+    `a=-0.75`), consumed by a fused `F.embedding_bag` or a weighted `embedding` sum. Fully vectorised
+    over packed patches (ragged `(h, w)` handled with `repeat_interleave`, no per-image loop), so it
+    supports dynamic shapes like the other grid_thw precompute helpers. `spatial_merge_size > 1`
+    emits patches in spatial-merge-block order (for encoders that consume merged tokens); `1` keeps
```
Comment (zucchini-nlp) [inline src/transformers/vision_utils.py:241]: nice!

File: src/transformers/exporters/utils.py (hunk 11 above)
Comment (IlyasMoutawwakil) [reply src/transformers/exporters/utils.py]: added !

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25, paddleocr_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25, paddleocr_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25, paddleocr_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25, paddleocr_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe

**Comment by IlyasMoutawwakil** [comment]
run-slow: kimi_k25, paddleocr_vl, qwen3_5, qwen3_5_moe, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe

**Comment by IlyasMoutawwakil** [comment]
yay
